libgadget/sfr_eff.c

```diff
@@ -359,24 +489,29 @@ starformation(int i, double *localsfr, double * sum_sm, double * sum_mass_stars)
     double prob = P[i].Mass / mass_of_star * (1 - exp(-p));
 
     if(get_random_number(P[i].ID + 1) < prob) {
-        *sum_mass_stars += mass_of_star;
-        retflag = make_particle_star(i, mass_of_star);
+        /* ok, make a star */
+        newstar = i;
+        /* If we get a fraction of the mass we need to create
+         * a new particle for the star and remove mass from i.*/
+        if(P[i].Mass >= 1.1 * mass_of_star)
+            newstar = slots_split_particle(i, mass_of_star);
     }
 
-    if(P[i].Type == 0)	{
-        /* to protect using a particle that has been turned into a star */
+    /* A particle that will be turned into a star doesn't want the metals or to be made a wind.*/
+    if(newstar != i)	{
```

Comment:
This line updates the metalicity of the SPH but does not update the metalicity if it has been turned into a star. I think with newstar != i, you are checking for 1. split happened? 2. no sfr happened? Probably cleaner if you (can) explicitly check for both.

Reply:
This was designed to match the behaviour before - if a star is formed it only includes a fraction of the metals added to the star-forming particle. Otherwise we can add the metals in one go a little earlier. I didn't want to change behaviour in this pull request - I don't know why it works like this though.

Reply:
Oh, I see what you were saying - yes that's clearer. I'll fix it.